packages/editor/src/store/actions.js

```diff
@@ -67,8 +175,76 @@ export function* setupEditor( post, edits, template ) {
 		blocks = synchronizeBlocksWithTemplate( blocks, template );
 	}
 
+	yield resetPost( post );
+	yield* resetLastBlockSourceDependencies();
+	yield {
+		type: 'SETUP_EDITOR',
+		post,
+		edits,
+		template,
+	};
 	yield resetEditorBlocks( blocks );
 	yield setupEditorState( post );
+	yield* __experimentalSubscribeSources();
+}
+
+/**
+ * Returns an action object signalling that the editor is being destroyed and
+ * that any necessary state or side-effect cleanup should occur.
+ *
+ * @return {Object} Action object.
+ */
+export function __experimentalTearDownEditor() {
+	return { type: 'TEAR_DOWN_EDITOR' };
+}
+
+/**
+ * Returns an action generator which loops to await the next state change,
+ * calling to reset blocks when a block source dependencies change.
+ *
+ * @yield {Object} Action object.
+ */
+export function* __experimentalSubscribeSources() {
+	while ( true ) {
+		yield awaitNextStateChange();
+
+		// The bailout case: If the editor becomes unmounted, it will flag
+		// itself as non-ready. Effectively unsubscribes from the registry.
+		const isStillReady = yield select( 'core/editor', '__unstableIsEditorReady' );
+		if ( ! isStillReady ) {
+			break;
+		}
+
+		const registry = yield getRegistry();
+
+		let reset = false;
+		for ( const source of Object.values( sources ) ) {
+			if ( ! source.getDependencies ) {
+				continue;
+			}
+
+			const dependencies = yield* source.getDependencies();
+
+			if ( ! lastBlockSourceDependenciesByRegistry.has( registry ) ) {
+				lastBlockSourceDependenciesByRegistry.set( registry, new WeakMap );
+			}
+
+			const lastBlockSourceDependencies = lastBlockSourceDependenciesByRegistry.get( registry );
```

Comment:
Can probably factor out this assignment and the optional `WeakMap` initialisation that comes before it.